tsl/src/fdw/estimate.c

```diff
@@ -285,53 +136,312 @@ adjust_foreign_grouping_path_cost(PlannerInfo *root, List *pathkeys, double retr
+			startup_cost += foreignrel->reltarget->cost.startup;
+			run_cost += foreignrel->reltarget->cost.per_tuple * rows;
+		}
+	}
+	else if (IS_JOIN_REL(foreignrel))
+	{
+		TsFdwRelInfo *fpinfo_i;
+		TsFdwRelInfo *fpinfo_o;
+		QualCost join_cost;
+		QualCost remote_conds_cost;
+		double nrows;
+
+		/* Use rows/width estimates made by the core code. */
+		rows = foreignrel->rows;
+		width = foreignrel->reltarget->width;
+
+		/* For join we expect inner and outer relations set */
+		Assert(fpinfo->innerrel && fpinfo->outerrel);
+
+		fpinfo_i = fdw_relinfo_get(fpinfo->innerrel);
+		fpinfo_o = fdw_relinfo_get(fpinfo->outerrel);
+
+		/* Estimate of number of rows in cross product */
+		nrows = fpinfo_i->rows * fpinfo_o->rows;
+
+		/*
+		 * Back into an estimate of the number of retrieved rows.  Just in
+		 * case this is nuts, clamp to at most nrows.
+		 */
+		retrieved_rows = clamp_row_est(rows / fpinfo->local_conds_sel);
+		retrieved_rows = Min(retrieved_rows, nrows);
+
+		/*
+		 * The cost of foreign join is estimated as cost of generating
+		 * rows for the joining relations + cost for applying quals on the
+		 * rows.
+		 */
+
+		/*
+		 * Calculate the cost of clauses pushed down to the foreign server
+		 */
+		cost_qual_eval(&remote_conds_cost, fpinfo->remote_conds, root);
+		/* Calculate the cost of applying join clauses */
+		cost_qual_eval(&join_cost, fpinfo->joinclauses, root);
+
+		/*
+		 * Startup cost includes startup cost of joining relations and the
+		 * startup cost for join and other clauses. We do not include the
+		 * startup cost specific to join strategy (e.g. setting up hash
+		 * tables) since we do not know what strategy the foreign server
+		 * is going to use.
+		 */
+		startup_cost = fpinfo_i->rel_startup_cost + fpinfo_o->rel_startup_cost;
+		startup_cost += join_cost.startup;
+		startup_cost += remote_conds_cost.startup;
+		startup_cost += fpinfo->local_conds_cost.startup;
+
+		/*
+		 * Run time cost includes:
+		 *
+		 * 1. Run time cost (total_cost - startup_cost) of relations being
+		 * joined
+		 *
+		 * 2. Run time cost of applying join clauses on the cross product
+		 * of the joining relations.
+		 *
+		 * 3. Run time cost of applying pushed down other clauses on the
+		 * result of join
+		 *
+		 * 4. Run time cost of applying nonpushable other clauses locally
+		 * on the result fetched from the foreign server.
+		 */
+		run_cost = fpinfo_i->rel_total_cost - fpinfo_i->rel_startup_cost;
+		run_cost += fpinfo_o->rel_total_cost - fpinfo_o->rel_startup_cost;
+		run_cost += nrows * join_cost.per_tuple;
+		nrows = clamp_row_est(nrows * fpinfo->joinclause_sel);
+		run_cost += nrows * remote_conds_cost.per_tuple;
+		run_cost += fpinfo->local_conds_cost.per_tuple * retrieved_rows;
+
+		/* Add in tlist eval cost for each output row */
+		startup_cost += foreignrel->reltarget->cost.startup;
+		run_cost += foreignrel->reltarget->cost.per_tuple * rows;
+	}
+	else if (IS_UPPER_REL(foreignrel))
+	{
```

Comment:
I would prefer that these changes are applied to `get_upper_rel_estimate()`? 

Now it is impossible to see the diff and whether changes we've made are lost or not. I don't think the goal here should be to 100% adopt PG's costing code as we've made changes and probably want to make further changes to adapt the code to account for async execution across multiple nodes.